devdoc/channel.md

```diff
@@ -0,0 +1,274 @@
+
+**SRS_CHANNEL_43_006: [** If the current `state` is `CHANNEL_PUSHED`, `channel_dispose` shall: **]**
+
+  - **SRS_CHANNEL_43_044: [** call the `push_callback` given to `channel_push` with `push_context` as the `push_context` given to `channel_push`, and `result` as `CHANNEL_RESULT_ABANDONED`. **]**
+
+  - **SRS_CHANNEL_43_045: [** release the reference to `data` stored in `channel` by calling `THANDLE_ASSIGN`. **]**
+
+### channel_pull
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_pull, THANDLE(CHANNEL), channel, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull);
+```
+
+`channel_pull` registers the given `pull_callback` to be called when there is data to be consumed.
+
+**SRS_CHANNEL_43_007: [** If `channel` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_008: [** If `pull_callback` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_009: [** If `out_op_pull` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_010: [** `channel_pull` shall call `interlocked_add` to obtain the `state` of the `channel`. **]**
+
+**SRS_CHANNEL_43_011: [** If the obtained `state` is `CHANNEL_PULLING` or `CHANNEL_PULLED`, `channel_pull` shall fail and return `CHANNEL_RESULT_REFUSED`. **]**
+
+**SRS_CHANNEL_43_059: [** If the obtained `state` is `CHANNEL_PUSHING`, `channel_pull` shall wait for the `state` to change by calling `InterlockedHL_WaitForNotValue`. **]**
+
+**SRS_CHANNEL_43_012: [** If the obtained `state` is `CHANNEL_PUSHED`, `channel_pull` shall : **]**
+
+  - **SRS_CHANNEL_43_060: [** store `push_callback`, `push_context` and `data` given to `channel_push`. **]**
+
+  - **SRS_CHANNEL_43_042: [** atomically compare the current `state` with the previously obtained `state`, increment the `EPOCH` and set the `state` to `CHANNEL_READY` if the `state` is has not changed by calling `interlocked_compare_exchange`. **]**
+
+  - **SRS_CHANNEL_43_071: [** If the `state` has changed, `channel_pull` shall retry from requirement 43_010. **]**
+
+  - **SRS_CHANNEL_43_013: [** call the given `pull_callback` with `pull_context` as the given `pull_context`, `data` as the stored `data` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+  - **SRS_CHANNEL_43_053: [** release the reference to the stored `data` by calling `THANDLE_ASSIGN`. **]**
+
+  - **SRS_CHANNEL_43_014: [** call the stored `push_callback` with the stored `push_context` and `result` as `CHANNEL_RESULT_OK`. **]**
+
+  - **SRS_CHANNEL_43_054: [** set `out_op_pull` to `NULL` by calling `THANDLE_ASSIGN`. **]**
+
+  - **SRS_CHANNEL_43_015: [** return `CHANNEL_RESULT_SYNC`. **]**
+
+**SRS_CHANNEL_43_016: [** If the obtained `state` is `CHANNEL_READY`, `channel_pull` shall: **]**
```

Comment:
>If the obtained `state` is `CHANNEL_READ [](http://example.com/codeflow?start=26&length=40)

I could not help but thinking about how we would use an event loop for this rather than baking interlocked operations. But just food for thought.